cns/ipampoolmonitor/ipampoolmonitor.go

```diff
@@ -11,121 +13,193 @@ import (
 )
 
 var (
-	increasePoolSize = 1
-	decreasePoolSize = -1
-	doNothing        = 0
+	increasePoolSize = int64(1)
+	decreasePoolSize = int64(-1)
+	doNothing        = int64(0)
 )
 
 type CNSIPAMPoolMonitor struct {
-	initialized bool
+	initialized    bool
+	pendingRelease bool
+
+	cachedSpec               nnc.NodeNetworkConfigSpec
+	cachedSecondaryIPConfigs map[string]cns.SecondaryIPConfig
 
 	cns            cns.HTTPService
 	rc             requestcontroller.RequestController
 	scalarUnits    cns.ScalarUnits
-	MinimumFreeIps int
-	MaximumFreeIps int
+	MinimumFreeIps int64
+	MaximumFreeIps int64
 
 	sync.RWMutex
 }
 
 func NewCNSIPAMPoolMonitor(cnsService cns.HTTPService, requestController requestcontroller.RequestController) *CNSIPAMPoolMonitor {
 	return &CNSIPAMPoolMonitor{
-		initialized: false,
-		cns:         cnsService,
-		rc:          requestController,
+		initialized:    false,
+		pendingRelease: false,
+		cns:            cnsService,
+		rc:             requestController,
 	}
 }
 
-// TODO: add looping and cancellation to this, and add to CNS MAIN
-func (pm *CNSIPAMPoolMonitor) Start() error {
-
-	if pm.initialized {
-		availableIPConfigs := pm.cns.GetAvailableIPConfigs()
-		rebatchAction := pm.checkForResize(len(availableIPConfigs))
-		switch rebatchAction {
-		case increasePoolSize:
-			return pm.increasePoolSize()
-		case decreasePoolSize:
-			return pm.decreasePoolSize()
-		}
+func stopReconcile(ch <-chan struct{}) bool {
+	select {
+	case <-ch:
+		return true
+	default:
 	}
 
-	return nil
+	return false
 }
 
-// UpdatePoolLimitsTransacted called by request controller on reconcile to set the batch size limits
-func (pm *CNSIPAMPoolMonitor) UpdatePoolLimitsTransacted(scalarUnits cns.ScalarUnits) {
-	pm.Lock()
-	defer pm.Unlock()
-	pm.scalarUnits = scalarUnits
+func (pm *CNSIPAMPoolMonitor) Start(poolMonitorRefreshMilliseconds int, exitChan <-chan struct{}) error {
+	logger.Printf("[ipam-pool-monitor] Starting CNS IPAM Pool Monitor")
+	for {
+		if stopReconcile(exitChan) {
+			return fmt.Errorf("CNS IPAM Pool Monitor received cancellation signal")
+		}
 
-	// TODO rounding?
-	pm.MinimumFreeIps = int(pm.scalarUnits.BatchSize * (pm.scalarUnits.RequestThresholdPercent / 100))
-	pm.MaximumFreeIps = int(pm.scalarUnits.BatchSize * (pm.scalarUnits.ReleaseThresholdPercent / 100))
+		err := pm.Reconcile()
+		if err != nil {
+			logger.Printf("[ipam-pool-monitor] CRITICAL %v", err)
```

Comment:
This is probably for offline discussion, but are there any cases where an error from reconciliation should prompt the app to crash and be restarted?